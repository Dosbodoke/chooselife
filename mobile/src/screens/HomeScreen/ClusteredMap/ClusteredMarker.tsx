import { View, Text } from 'react-native';
import { Marker } from 'react-native-maps';

interface Props {
  coordinate: { latitude: number; longitude: number };
  pointCount: number;
  size: number;
}

const ClusteredMarker = (props: Props) => {
  return (
    <Marker coordinate={props.coordinate}>
      <View
        className="rounded-full bg-slate-700 flex items-center justify-center"
        style={{ width: props.size, height: props.size }}>
        <Text className="text-white text-center font-bold text-xl">{props.pointCount}</Text>
      </View>
    </Marker>
  );
};

export default ClusteredMarker;
