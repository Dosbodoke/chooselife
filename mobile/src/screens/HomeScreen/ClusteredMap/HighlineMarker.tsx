import type { Coordinates } from '@src/database';
import { useAppDispatch } from '@src/redux/hooks';
import { highliteMarker, minimizeMarker } from '@src/redux/slices/markerSlice';
import { Marker, Polyline } from 'react-native-maps';

interface Props {
  id: string;
  isHighlited?: boolean;
  coordinateA: Coordinates;
  coordinateB: Coordinates;
}

const HighlineMarker = (props: Props) => {
  const dispatch = useAppDispatch();

  const handleOnPress = () => {
    if (props.isHighlited) {
      dispatch(minimizeMarker());
      return;
    }
    dispatch(
      highliteMarker({
        type: 'Highline',
        id: props.id,
        coords: [props.coordinateA, props.coordinateB],
      })
    );
  };

  return (
    <>
      <Marker coordinate={props.coordinateA} onPress={handleOnPress} />
      {props.isHighlited && (
        <>
          <Marker coordinate={props.coordinateB} onPress={handleOnPress} />
          <Polyline coordinates={[props.coordinateA, props.coordinateB]} strokeWidth={3} />
        </>
      )}
    </>
  );
};

export default HighlineMarker;
