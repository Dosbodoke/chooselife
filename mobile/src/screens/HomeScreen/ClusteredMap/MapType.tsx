import { useNavigation } from '@react-navigation/native';
import { MapTypeSvg } from '@src/assets';
import type { HomeScreenProps } from '@src/navigation/types';
import { View, TouchableOpacity } from 'react-native';

type NavigationProp = HomeScreenProps['navigation'];

interface Props {
  mBottom: number;
}

const MapType = ({ mBottom }: Props) => {
  const navigation = useNavigation<NavigationProp>();

  return (
    <View className="absolute right-2 mb-16" style={{ bottom: mBottom }}>
      <TouchableOpacity
        className="h-10 w-10 rounded-full bg-gray-100 justify-center items-center"
        onPress={() => {
          navigation.navigate('MapType');
        }}>
        <MapTypeSvg />
      </TouchableOpacity>
    </View>
  );
};

export default MapType;
