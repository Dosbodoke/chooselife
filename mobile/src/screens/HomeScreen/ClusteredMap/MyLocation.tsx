import { MyLocationSvg } from '@src/assets';
import { View, TouchableOpacity } from 'react-native';

interface Props {
  mBottom: number;
  onPress: () => void;
}

const MyLocation = ({ onPress, mBottom }: Props) => {
  return (
    <View className="absolute right-2 mb-3" style={{ bottom: mBottom }}>
      <TouchableOpacity
        className="h-10 w-10 rounded-full bg-gray-100 justify-center items-center"
        onPress={onPress}>
        <MyLocationSvg />
      </TouchableOpacity>
    </View>
  );
};

export default MyLocation;
