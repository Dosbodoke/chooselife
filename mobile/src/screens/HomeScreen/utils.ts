import { BBox } from 'geojson';
import { Region } from 'react-native-maps';

export const regionToBoundingBox = (region: Region): BBox => {
  let lngD: number;
  if (region.longitudeDelta < 0) lngD = region.longitudeDelta + 360;
  else lngD = region.longitudeDelta;

  return [
    region.longitude - lngD, // westLng - min lng
    region.latitude - region.latitudeDelta, // southLat - min lat
    region.longitude + lngD, // eastLng - max lng
    region.latitude + region.latitudeDelta, // northLat - max lat
  ];
};

export const toRad = (number: number) => (number * Math.PI) / 180;

export const toDeg = (rad: number) => rad * (180 / Math.PI);
