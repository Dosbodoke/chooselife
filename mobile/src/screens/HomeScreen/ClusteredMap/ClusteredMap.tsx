import { MIN_MARKER_SIZE } from '@src/constants';
import type { Coordinates } from '@src/database';
import database from '@src/database';
import { useAppSelector } from '@src/redux/hooks';
import { selectHighlitedMarker, selectMapType } from '@src/redux/slices/mapSlice';
import { BBox, GeoJsonProperties } from 'geojson';
import { useState, useRef, useMemo, useEffect } from 'react';
import MapView, { Region } from 'react-native-maps';
import { PointFeature } from 'supercluster';
import useSuperCluster from 'use-supercluster';

import { regionToBoundingBox, getMyLocation } from '../utils';
import ClusteredMarker from './ClusteredMarker';
import HighlineMarker from './HighlineMarker';
import MapType from './MapType';
import MyLocation from './MyLocation';

interface PointProperties {
  cluster: boolean;
  category: string;
  highId: string;
  anchorB: Coordinates;
}

interface Props {
  buttonMarginBottom: number;
}

const ClusteredMap = ({ buttonMarginBottom }: Props) => {
  const initialRegion = {
    latitude: -15.7782081,
    longitude: -47.93371,
    latitudeDelta: 5,
    longitudeDelta: 5,
  };

  const mapRef = useRef<MapView>(null);
  const highlitedMarker = useAppSelector(selectHighlitedMarker);
  const mapType = useAppSelector(selectMapType);
  const [bounds, setBounds] = useState<BBox>(regionToBoundingBox(initialRegion));
  const [zoom, setZoom] = useState(10);

  const setMyLocation = async () => {
    const region = await getMyLocation();
    region && mapRef.current?.animateToRegion(region, 1000);
  };

  const onRegionChangeComplete = async (region: Region) => {
    const mapBound = regionToBoundingBox(region);
    const coords = await mapRef.current?.getCamera();
    setZoom(coords?.zoom ?? 10);
    setBounds(mapBound);
  };

  const fitMapToCoords = (coords: Coordinates[]): void => {
    mapRef.current?.fitToCoordinates(coords, {
      edgePadding: {
        top: 0,
        right: 50,
        bottom: 100,
        left: 50,
      },
      animated: true,
    });
  };

  const points = useMemo<PointFeature<GeoJsonProperties & PointProperties>[]>(() => {
    return database?.highline.map((h) => ({
      type: 'Feature',
      properties: {
        cluster: false,
        category: 'highline',
        highId: h.id,
        anchorB: h.anchorB,
      },
      geometry: {
        type: 'Point',
        coordinates: [h.anchorA.longitude, h.anchorA.latitude],
      },
    }));
  }, [database?.highline]);

  const { clusters } = useSuperCluster({
    points,
    bounds,
    zoom,
    options: { radius: 75, maxZoom: 25 },
  });

  useEffect(() => {
    highlitedMarker && fitMapToCoords(highlitedMarker.coords);
  }, [highlitedMarker]);

  return (
    <>
      <MapView
        className="flex-1"
        testID="MapView"
        provider="google"
        mapType={mapType}
        initialRegion={initialRegion}
        ref={mapRef}
        onRegionChangeComplete={onRegionChangeComplete}>
        {highlitedMarker && (
          <HighlineMarker
            id={highlitedMarker.id}
            coordinateA={highlitedMarker.coords[0]}
            coordinateB={highlitedMarker.coords[1]}
            isHighlited
          />
        )}
        {clusters?.map((point) => {
          const [longitude, latitude] = point.geometry.coordinates;
          if (typeof longitude !== 'number' || typeof latitude !== 'number') return;
          const coordinateA = { latitude, longitude };
          const properties = point.properties;

          if (properties?.cluster) {
            const size = Math.max((properties.point_count * 40) / points.length, MIN_MARKER_SIZE);
            return (
              <ClusteredMarker
                key={`cluster-${properties.cluster_id}`}
                size={size}
                coordinate={coordinateA}
                pointCount={properties.point_count}
              />
            );
          }

          const coordinateB = {
            latitude: properties.anchorB.latitude,
            longitude: properties.anchorB.longitude,
          };
          return (
            <HighlineMarker
              key={`marker-high-${properties.highId}`}
              id={properties.highId}
              coordinateA={coordinateA}
              coordinateB={coordinateB}
            />
          );
        })}
      </MapView>
      <MyLocation mBottom={buttonMarginBottom} onPress={setMyLocation} />
      <MapType mBottom={buttonMarginBottom} />
    </>
  );
};

export default ClusteredMap;
