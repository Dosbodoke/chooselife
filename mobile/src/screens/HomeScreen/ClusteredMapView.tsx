import * as Location from 'expo-location';
import { BBox, GeoJsonProperties } from 'geojson';
import { useState, useRef } from 'react';
import MapView, { Marker, Region } from 'react-native-maps';
import { PointFeature } from 'supercluster';
import useSuperCluster from 'use-supercluster';

import database from '../../database';
import ClusteredMarker from './ClusteredMarker';
import MyLocation from './MyLocation';
import { regionToBoundingBox } from './utils';

interface Props {
  buttonMarginBottom: number;
}

const getMyLocation = async (): Promise<Region | undefined> => {
  const { status } = await Location.requestForegroundPermissionsAsync();
  if (status !== 'granted') return;

  const { latitude, longitude } = (await Location.getCurrentPositionAsync({})).coords;
  const region = {
    latitude,
    longitude,
    latitudeDelta: 0.035,
    longitudeDelta: 0.035,
  };
  return region;
};

const minMarkerSize = 30;

const ClusteredMapView = ({ buttonMarginBottom }: Props) => {
  const initialRegion = {
    latitude: -15.7782081,
    longitude: -47.93371,
    latitudeDelta: 5,
    longitudeDelta: 5,
  };
  const initialBound = regionToBoundingBox(initialRegion);
  const [bounds, setBounds] = useState<BBox>(initialBound);
  const [zoom, setZoom] = useState(10);

  const mapRef = useRef<MapView>(null);

  const setMyLocation = async () => {
    const region = await getMyLocation();
    region && mapRef.current?.animateToRegion(region, 1000);
  };

  const onRegionChangeComplete = async (region: Region) => {
    const mapBound = regionToBoundingBox(region);
    const coords = await mapRef.current?.getCamera();
    setZoom(coords?.zoom ?? 10);
    setBounds(mapBound);
  };

  const points: PointFeature<GeoJsonProperties>[] = database?.highline.map((h) => ({
    type: 'Feature',
    properties: {
      cluster: false,
      category: 'highline',
      highId: h.id,
    },
    geometry: {
      type: 'Point',
      coordinates: [h.anchorA.longitude, h.anchorA.latitude],
    },
  }));

  const { clusters } = useSuperCluster({
    points,
    bounds,
    zoom,
    options: { radius: 75, maxZoom: 25 },
  });

  return (
    <>
      <MapView
        className="flex-1"
        testID="MapView"
        provider="google"
        initialRegion={initialRegion}
        ref={mapRef}
        onRegionChangeComplete={onRegionChangeComplete}
        showsUserLocation>
        {clusters?.map((point) => {
          const [longitude, latitude] = point.geometry.coordinates;
          const coordinate = { latitude, longitude };
          const properties = point.properties ?? {};

          if (properties?.cluster) {
            const size = Math.max((properties.point_count * 40) / points.length, minMarkerSize);
            return (
              <ClusteredMarker
                key={`cluster-${properties.cluster_id}`}
                size={size}
                coordinate={coordinate}
                pointCount={properties.point_count}
              />
            );
          }
          return <Marker key={`marker-${properties.highId}`} coordinate={coordinate} />;
        })}
      </MapView>
      <MyLocation mBottom={buttonMarginBottom} onPress={setMyLocation} />
    </>
  );
};

export default ClusteredMapView;
