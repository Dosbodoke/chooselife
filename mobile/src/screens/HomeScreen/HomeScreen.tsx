import { useAppSelector } from '@src/redux/hooks';
import { selectHighlitedMarker } from '@src/redux/slices/mapSlice';
import { useState } from 'react';
import { View, StatusBar, LayoutChangeEvent } from 'react-native';

import type { HomeScreenProps } from '../../navigation/types';
import ClusteredMap from './ClusteredMap/ClusteredMap';
import DetailCard from './DetailCard/DetailCard';
import SearchCard from './SearchCard/SearchCard';

const HomeScreen = ({ navigation }: HomeScreenProps) => {
  const [marginBottom, setMarginBottom] = useState(0);
  const highlitedMarker = useAppSelector(selectHighlitedMarker);

  const handleLayoutChange = (event: LayoutChangeEvent) => {
    const { height } = event.nativeEvent.layout;
    setMarginBottom(height);
  };

  return (
    <View className="relative h-full">
      <StatusBar barStyle="dark-content" />

      <ClusteredMap buttonMarginBottom={marginBottom} />

      {highlitedMarker ? (
        <DetailCard handleLayoutChange={handleLayoutChange} highlitedMarker={highlitedMarker} />
      ) : (
        <SearchCard handleLayoutChange={handleLayoutChange} navigation={navigation} />
      )}
    </View>
  );
};

export default HomeScreen;
