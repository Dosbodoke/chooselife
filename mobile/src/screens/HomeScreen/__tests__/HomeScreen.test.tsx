/* eslint-disable @typescript-eslint/no-var-requires */
/* eslint-disable @typescript-eslint/no-explicit-any */
/* eslint-disable react/display-name */

import { renderWithProviders, createTestProps } from '@src/utils/test-utils';

import ClusteredMap from '../ClusteredMap/ClusteredMap';
import HomeScreen from '../HomeScreen';

jest.mock('react-native/Libraries/Animated/NativeAnimatedHelper');

jest.mock('react-native-maps', () => {
  const React = require('react');
  const { View } = require('react-native');

  const MockMapView = React.forwardRef(({ testID, children, ...props }: any, ref: any) => {
    if (ref?.current) {
      ref.current = {
        getMapBoundaries: async () => ({
          northEast: {
            latitude: 2,
            longitude: 2,
          },
          southWest: {
            latitude: 1,
            longitude: 1,
          },
        }),
        getCamera: async () => ({
          center: {
            latitude: 2,
            longitude: 2,
          },
          heading: 1,
          pitch: 1,
          zoom: 1,
          altitude: 1,
        }),
        animateCamera: () => {},
      };
    }

    return (
      <View testID={testID} {...props}>
        {children}
      </View>
    );
  });

  const MockMarker = React.forwardRef(({ testID, children, ...props }: any, ref: any) => {
    if (ref?.current) {
      ref.current = {
        redraw: () => {},
      };
    }

    return (
      <View testID={testID} {...props}>
        {children}
      </View>
    );
  });

  const mockMapTypes = {
    STANDARD: 0,
    SATELLITE: 1,
    HYBRID: 2,
    TERRAIN: 3,
    NONE: 4,
    MUTEDSTANDARD: 5,
  };

  return {
    __esModule: true,
    default: MockMapView,
    Marker: MockMarker,
    MAP_TYPES: mockMapTypes,
    PROVIDER_DEFAULT: 'default',
    PROVIDER_GOOGLE: 'google',
  };
});

describe('HomeScreen', () => {
  describe('Map', () => {
    it('Match snapshot', () => {
      // when
      const tree = renderWithProviders(<ClusteredMap buttonMarginBottom={240} />).toJSON();

      // then
      expect(tree).toMatchSnapshot();
    });

    it('Render the map', () => {
      // given
      const props: any = createTestProps({});

      // when
      const { getByTestId } = renderWithProviders(<HomeScreen {...props} />);
      const map = getByTestId('MapView');

      // then
      expect(map).toBeDefined();
    });
  });
});
