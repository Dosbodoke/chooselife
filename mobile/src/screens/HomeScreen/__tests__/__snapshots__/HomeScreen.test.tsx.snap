// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HomeScreen Map Renders correctly 1`] = `
Array [
  <View
    initialRegion={
      Object {
        "latitude": -15.7782081,
        "latitudeDelta": 5,
        "longitude": -47.93371,
        "longitudeDelta": 5,
      }
    }
    onRegionChangeComplete={[Function]}
    provider="google"
    showsUserLocation={true}
    style={
      Array [
        Object {
          "flexBasis": "0%",
          "flexGrow": 1,
          "flexShrink": 1,
        },
      ]
    }
    testID="MapView"
  />,
  <View
    style={
      Array [
        Object {
          "position": "absolute",
        },
        Object {
          "right": 8,
        },
        Object {
          "bottom": undefined,
        },
      ]
    }
  >
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#f3f4f6",
          "borderBottomLeftRadius": 9999,
          "borderBottomRightRadius": 9999,
          "borderTopLeftRadius": 9999,
          "borderTopRightRadius": 9999,
          "height": 40,
          "justifyContent": "center",
          "opacity": 1,
          "width": 40,
        }
      }
    >
      <RNSVGSvgView
        align="xMidYMid"
        bbHeight="24px"
        bbWidth="24px"
        focusable={false}
        height="24px"
        meetOrSlice={0}
        minX={0}
        minY={0}
        style={
          Array [
            Object {
              "backgroundColor": "transparent",
              "borderWidth": 0,
            },
            Object {
              "flex": 0,
              "height": 24,
              "width": 24,
            },
          ]
        }
        vbHeight={24}
        vbWidth={24}
        width="24px"
      >
        <RNSVGGroup>
          <RNSVGPath
            d="M12 2a.75.75 0 01.743.648l.007.102v1.788a7.5 7.5 0 016.713 6.715l.037-.003h1.75a.75.75 0 01.102 1.493l-.102.007-1.788-.001a7.5 7.5 0 01-6.715 6.714l.003.037v1.75a.75.75 0 01-1.493.102l-.007-.102.001-1.788a7.5 7.5 0 01-6.714-6.715l-.037.003H2.75a.75.75 0 01-.102-1.493l.102-.007h1.788a7.5 7.5 0 016.715-6.713L11.25 4.5V2.75A.75.75 0 0112 2zm0 4a6 6 0 100 12 6 6 0 000-12zm0 2a4 4 0 110 8 4 4 0 010-8z"
            fill={4280361249}
            fillRule={1}
            propList={
              Array [
                "fill",
                "fillRule",
                "stroke",
                "strokeWidth",
              ]
            }
            stroke={null}
            strokeWidth={1}
          />
        </RNSVGGroup>
      </RNSVGSvgView>
    </View>
  </View>,
]
`;
